Source/JavaScriptCore/dfg/DFGInPlaceAbstractState.cpp

```diff
@@ -421,14 +418,289 @@ inline bool InPlaceAbstractState::mergeToSuccessors(BasicBlock* basicBlock)
+                    break;
+                AbstractValue trueValue = forNode(edge);
+                AbstractValue falseValue = trueValue;
+                if (trueValue.filter(SpecObject) == FiltrationOK)
+                    trueNodes.append({ edge.node(), trueValue });
+                if (falseValue.filter(SpecOther) == FiltrationOK)
+                    falseNodes.append({ edge.node(), falseValue });
+                break;
+            }
+            case BooleanUse:
+            case KnownBooleanUse:
+                falseNodes.append({ edge.node(), constant(edge.node(), jsBoolean(false)) });
+                trueNodes.append({ edge.node(), constant(edge.node(), jsBoolean(true)) });
+
+                switch (edge->op()) {
+                case CompareEq:
+                case CompareStrictEq: {
+                    Node* compare = edge.node();
+
+                    bool isValidCompare = [&]() {
+                        if (compare->op() == CompareStrictEq)
+                            return true;
+                        return compare->isBinaryUseKind(Int32Use)
+#if USE(JSVALUE64)
+                            || compare->isBinaryUseKind(Int52RepUse)
+#endif
+                            || compare->isBinaryUseKind(DoubleRepUse)
+                            || compare->isBinaryUseKind(StringUse)
+                            || compare->isBinaryUseKind(StringIdentUse)
+                            || compare->isBinaryUseKind(BooleanUse)
+                            || compare->isBinaryUseKind(ObjectUse)
+                            || compare->isBinaryUseKind(SymbolUse);
+                    }();
+                    if (!isValidCompare)
+                        break;
+
+                    AbstractValue lhs = forNode(compare->child1());
+                    AbstractValue rhs = forNode(compare->child2());
+                    if ((lhs.m_type & SpecDoubleNaN) || (rhs.m_type & SpecDoubleNaN)) {
+                        // We can't say anything useful about this comparison if either
+                        // input could be NaN, since it doesn't tell us any information
+                        // about the underlying values.
+                        break;
+                    }
+
+                    if (lhs.filter(rhs) == FiltrationOK) {
+                        if (JSValue value = lhs.value()) {
+                            ASSERT(!(value.isDouble() && std::isnan(value.asDouble()))); // This is handled by the above NaN type check.
+                            // 0.0 == -0.0, so we can't have this filter produce a constant
+                            // if that constant is zero or negative zero.
+                            if (value.isDouble() && value.asNumber() == 0.0)
+                                lhs.clobberValue();
+                        }
+                        trueNodes.append({ compare->child1().node(), lhs });
+                        trueNodes.append({ compare->child2().node(), lhs });
```

Comment:
This seems not correct, right? Let's say `0 === -0` or `0 == -0` which returns true. Then, we can say two childs' true nodes correspond to the same abstract value of the left child (child1). This is because `0` and `-0` are still different, right?
```
1 / 0 // Infinity
1 / -0 // -Infinity
```

Reply:
hmm, maybe it's fine? Since `-0` must be a double and at this point both should be in double type.

Reply:
Yeah, I think this is fine. So long as it is not constant, then 0 / -0 does not have difference in AbstractValue's types.